### ISSUE TITLE
append! fails

### ISSUE THREAD

**Issue by kafisatz** [JuliaData/SentinelArrays.jl#52]
I have a loop calling `df = CSV.read(...)` and then `append!(bigdf,df)`.
I end up with different types for one of the columns which append! cannot process.

Should `append!` be able to manage this?
Or should I add additional code to 'get rid of' the PooledArray / ChainedVector? 

```
using DataFrames
using DataFrames.PooledArrays
using CSV
using CSV.SentinelArrays

mv2 = convert(Vector{Union{Missing, String}},MissingVector(9))
pa2 = PooledArray(mv2,UInt32)
df = DataFrame(a=rand(9),b=pa2)
cv = ChainedVector([MissingVector(999) for i = 1:12])
df2 = DataFrame(a=rand(999*12),b=cv)
df2.b = cv
append!(df,df2,promote=true)



julia> append!(df,df2,promote=true)
┌ Error: Error adding value to column :b.
└ @ DataFrames C:\Users\bernhard.konig.ROOT_MILLIMAN\.julia\packages\DataFrames\vQokV\src\dataframe\dataframe.jl:1335
ERROR: MethodError: no method matching copyto!(::PooledArray{Union{Missing, String},UInt32,1,Array{UInt32,1}}, ::Int64, ::ChainedVector{Missing,MissingVector}, ::SentinelArrays.ChainedVectorIndex{MissingVector}, ::Int64)
Closest candidates are:
  copyto!(::AbstractArray{T,1}, ::Union{Signed, Unsigned}, ::ChainedVector, ::Union{Signed, Unsigned}, ::Union{Signed, Unsigned}) where T at C:\Users\bernhard.konig.ROOT_MILLIMAN\.julia\packages\SentinelArrays\cS7Oh\src\chainedvector.jl:325
  copyto!(::AbstractArray, ::Integer, ::AbstractArray, ::Integer, ::Integer) at abstractarray.jl:899
  copyto!(::AbstractArray, ::Integer, ::Any, ::Integer, ::Integer) at abstractarray.jl:779
  ...
Stacktrace:
 [1] append!(::PooledArray{Union{Missing, String},UInt32,1,Array{UInt32,1}}, ::ChainedVector{Missing,MissingVector}) at C:\Users\bernhard.konig.ROOT_MILLIMAN\.julia\packages\PooledArrays\CV8kA\src\PooledArrays.jl:534
 [2] append!(::DataFrame, ::DataFrame; cols::Symbol, promote::Bool) at C:\Users\bernhard.konig.ROOT_MILLIMAN\.julia\packages\DataFrames\vQokV\src\dataframe\dataframe.jl:1290
 [3] top-level scope at REPL[11]:1


#typeof(df.b) = PooledArray{Union{Missing, String},UInt32,1,Array{UInt32,1}}
#typeof(df2.b) = ChainedVector{Missing,MissingVector}


(@v1.5) pkg> st
Status `C:\Users\bernhard.konig.ROOT_MILLIMAN\.julia\environments\v1.5\Project.toml`
  [336ed68f] CSV v0.8.5
  [a93c6f00] DataFrames v1.2.0
  [ffa9a821] DocumentFormat v3.2.3
  [48062228] FilePathsBase v0.9.10
  [4138dd39] JLD v0.12.1
  [14b8a8f1] PkgTemplates v0.7.15
  [c46f51b8] ProfileView v0.6.9

julia> versioninfo()
Julia Version 1.5.4
Commit 69fcb5745b (2021-03-11 19:13 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake-avx512)
Environment:
  JULIA_EDITOR = "C:\Program Files\Microsoft VS Code\Code.exe"
  JULIA_NUM_THREADS = 16

julia>
```

**Comment by bkamins** [comment]
Ah - moving it again. It is an issue with SentinelArrays.jl.

@quinnj - the problem is that `eachindex` in sentinel array does produce custom indexing which does not play well with `copyto!` definition in Julia Base.

So either:
1. `eachindex` should produce a more standard value (i.e. integer), but I guess it is going to be inefficient
2. `copyto!` with `ChainedVector` as source or target get a custom implementation (this can be tricky and lead to method ambiguities)
3. In PooledArrays.jl we should not use `copyto!` the way we do now (but this is a local fix and I assume this issue will re-emerge in other situations then)

**Comment by kafisatz** [comment]
Thanks. I am glad I am not the only one being confused with `PooledArrays, SentinelArrays, ChainedVectors` and the likes; although I am sure these all have their purpose.

**Comment by bkamins** [comment]
I just needed to pinpoint the root cause :). As you can see here:
```
julia> x = ChainedVector([[10,11],[12],[13,14,15]])
6-element ChainedVector{Int64, Vector{Int64}}:
 10
 11
 12
 13
 14
 15

julia> for i in eachindex(x)
       @show i
       @show x[i]
       end
i = SentinelArrays.ChainedVectorIndex{Vector{Int64}}([10, 11], 1)
x[i] = 10
i = SentinelArrays.ChainedVectorIndex{Vector{Int64}}([10, 11], 2)
x[i] = 11
i = SentinelArrays.ChainedVectorIndex{Vector{Int64}}([12], 1)
x[i] = 12
i = SentinelArrays.ChainedVectorIndex{Vector{Int64}}([13, 14, 15], 1)
x[i] = 13
i = SentinelArrays.ChainedVectorIndex{Vector{Int64}}([13, 14, 15], 2)
x[i] = 14
i = SentinelArrays.ChainedVectorIndex{Vector{Int64}}([13, 14, 15], 3)
x[i] = 15
```
`ChainedVector` uses a custom index type for efficiency reasons. And because of this it does not integrate well with Julia Base, which assumes that indexing is either by integers or by `CartesianIndices` in functions like `copyto!`.

**Comment by bkamins** [comment]
@quinnj - a potentially best solution would be to make `SentinelArrays.ChainedVectorIndex <: Integer` and for an "outside world" it would be just a normal 1-based integer index (with a subset of `Integer` operations supported - a minimum required is probably conversion to other integer types), but for the SentinelArrays.jl it could be given a special treatment.

I think it would be cleanest solution if this would be feasible.

**Comment by bkamins** [comment]
@quinnj + @nalimilan also:
```
julia> using SentinelArrays, CategoricalArrays

julia> x = ChainedVector([[1, 2, 3, 4], [5, 6, 7, 8]])
8-element ChainedVector{Int64, Vector{Int64}}:
 1
 2
 3
 4
 5
 6
 7
 8

julia> cut(x, 5)
ERROR: ArgumentError: invalid index: SentinelArrays.ChainedVectorIndex{Vector{Int64}}([1, 2, 3, 4], 1) of type SentinelArrays.ChainedVectorIndex{Vector{Int64}}

julia> cut(copy(x), 5)
8-element CategoricalArray{String,1,UInt32}:
 "Q1: [1.0, 2.4000000000000004)"
 "Q1: [1.0, 2.4000000000000004)"
 "Q2: [2.4000000000000004, 3.8000000000000003)"
 "Q3: [3.8000000000000003, 5.2)"
 "Q3: [3.8000000000000003, 5.2)"
 "Q4: [5.2, 6.6000000000000005)"
 "Q5: [6.6000000000000005, 8.0]"
 "Q5: [6.6000000000000005, 8.0]"
```

**Comment by quinnj** [comment]
Fixed in #53
